llama_index/chat_engine/cohere_context_plus_citations.py

```diff
@@ -0,0 +1,170 @@
+import asyncio
+from threading import Thread
+from typing import Any, List, Optional, Tuple, Dict
+
+from llama_index.callbacks import trace_method
+from llama_index.chat_engine.types import (
+    AgentChatResponse,
+    StreamingAgentChatResponse,
+    ToolOutput,
+)
+from llama_index.core.base_retriever import BaseRetriever
+from llama_index.core.llms.types import ChatMessage, MessageRole
+from llama_index.llms.llm import LLM
+from llama_index.schema import NodeWithScore
+from llama_index.chat_engine import ContextChatEngine
+from llama_index.llms.cohere_utils import transform_nodes_to_cohere_documents_list
+
+
+class CohereContextPlusCitationsChatEngine(ContextChatEngine):
+    """Cohere Context + Citations Chat Engine.
+
+    Uses a retriever to retrieve a context, set the context in the Cohere
+    documents param(https://docs.cohere.com/docs/retrieval-augmented-generation-rag),
+    and then uses an LLM to generate a response.
+    """
+
+    @trace_method("chat")
+    def chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> AgentChatResponse:
+        if chat_history is not None:
+            self._memory.set(chat_history)
+        self._memory.put(ChatMessage(content=message, role="user"))
+        context_str_template, nodes = self._generate_context(message)
+        prefix_messages = self._get_prefix_messages_with_context(context_str_template)
+
+        all_messages = self._memory.get_all()
+        # transform nodes to cohere documents list
+        documents_list = transform_nodes_to_cohere_documents_list(nodes)
+        # and then uses an LLM to generate a response
+        chat_response = self._llm.chat(all_messages, documents=documents_list)
+        # TODO ask TJ do we need it?
+        ai_message = chat_response.message
+        self._memory.put(ai_message)
```

Comment:
Should the `ai_message` on line 44 be a ChatMessage object?

Reply:
Yes - it's a ChatMessage object